### PR TITLE
os/bluestore/BlueFS: reduce bufferlist rebuild

### PR DESCRIPTION
**Pull request by majianpeng**
If bufferlist use class page_aligned_appender, this make
bufferlist::append_zero can't append data into
page_aligned_append::buffer and it will create a new ptr into
bufferlist.
This make bufferlist rebuild. This patch don't remove rebuild this
because page_aligned_appender don't use bufferlist::_carriage. This
patch split the size_unalignd ptr into two parts: first part is aligned
and second is unaligned which size is less than 4k.
This make only 4k rebuild.

Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by majianpeng** [comment]
@tchaikov .I firstly try modify class page_aligned_appender but it's a difficulty. Luckily only bluefs use this class. So i fix the useage.

**Comment by majianpeng** [comment]
@tchaikov . ping .

**Comment by majianpeng** [comment]
@rzarzynski . i rework the split process. But i don't know whether better than the previous implement. please review.

**Comment by majianpeng** [comment]
@rzarzynski . I create a new PR which print the memcopy_count.  https://github.com/ceph/ceph/pull/36514

**Comment by rzarzynski** [comment]
@majianpeng: OK, I'm going to take your commit and (try to) rework `page_aligned_appender` on top of it.

**Comment by majianpeng** [comment]
@rzarzynski @tchaikov . Whether to consider merging first this PR?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-08-28_06:18:09-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-28_13:09:24-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-08-29_04:43:21-rados-wip-kefu-testing-2020-08-29-1106-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-29_13:54:19-rados-wip-kefu-testing-2020-08-29-1945-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-31_03:26:38-rados-wip-kefu-testing-2020-08-31-0934-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/47179

**Review by ifed01 — APPROVED** [review]
LGTM

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
@tchaikov .  update. please review . Thanks!

**Comment by majianpeng** [comment]
retest this please

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -2592,6 +2592,16 @@ int BlueFS::_flush_range(FileWriter *h, uint64_t offset, uint64_t length)
              << " and padding block with 0x" << (super.block_size - tail)
              << std::dec << dendl;
     h->tail_block.substr_of(bl, bl.length() - tail, tail);
+    //Because class page_aligned_appender, bl.append_zero will create a new ptr.
```
Comment (tchaikov) [inline src/os/bluestore/BlueFS.cc:2595]: please add a space after `//`.

**Comment by tchaikov** [comment]
@majianpeng i since @rzarzynski @ifed01 have reviewed and approved your change. i guess this change is good to go after `[RFC]` is removed from the title?

**Comment by majianpeng** [comment]
@tchaikov .   updated by your two suggestions(remove RFC and  space after //)

**Comment by tchaikov** [comment]
closing in favor of #36754

**Pull request closed without merging**